pkg/controllers/node/node_syncer.go

```diff
@@ -45,42 +59,71 @@ func (c *NodeController) OnUpdates(updates []bapi.Update) {
 			fallthrough
+
 		case bapi.UpdateTypeKVUpdated:
 			n := upd.KVPair.Value.(*apiv3.Node)
-			if kn := getK8sNodeName(*n); kn != "" {
-				// Create a mapping from Kubernetes node -> Calico node.
-				logrus.Debugf("Mapping k8s node -> calico node. %s -> %s", kn, n.Name)
-				c.nodemapLock.Lock()
-				c.nodemapper[kn] = n.Name
-				c.nodemapLock.Unlock()
-
-				// It has a node reference - get that Kubernetes node, and if
-				// it exists perform a sync.
-				obj, ok, err := c.indexer.GetByKey(kn)
-				if !ok {
-					logrus.Debugf("No corresponding kubernetes node")
-					continue
-				} else if err != nil {
-					logrus.WithError(err).Warnf("Couldn't get node from indexer")
-					continue
+
+			if c.config.SyncNodeLabels && c.config.DatastoreType != "kubernetes" {
+				if kn := getK8sNodeName(*n); kn != "" {
+					// Create a mapping from Kubernetes node -> Calico node.
+					logrus.Debugf("Mapping k8s node -> calico node. %s -> %s", kn, n.Name)
+					c.nodemapLock.Lock()
+					c.nodemapper[kn] = n.Name
+					c.nodemapLock.Unlock()
+
+					// It has a node reference - get that Kubernetes node, and if
+					// it exists perform a sync.
+					obj, ok, err := c.indexer.GetByKey(kn)
+					if !ok {
+						logrus.Debugf("No corresponding kubernetes node")
+						continue
+					} else if err != nil {
+						logrus.WithError(err).Warnf("Couldn't get node from indexer")
+						continue
+					}
+					c.syncNodeLabels(obj.(*v1.Node))
+				}
+			}
+
+			if c.config.AutoHostEndpoints == "enabled" {
+				// Cache all updated nodes.
+				c.nodeCache[n.Name] = n
+
+				// If we're already in-sync, sync the node's auto hostendpoint.
+				if c.syncStatus == bapi.InSync {
+					err := c.syncAutoHostendpointWithRetries(n)
+					if err != nil {
+						logrus.WithError(err).Fatal()
+					}
 				}
-				c.syncNodeLabels(obj.(*v1.Node))
 			}
+
 		case bapi.UpdateTypeKVDeleted:
 			if upd.KVPair.Value != nil {
 				logrus.Warnf("KVPair value should be nil for Deleted UpdataType")
 			}
 
+			nodeName := upd.KVPair.Key.(model.ResourceKey).Name
+
 			// Try to perform unmapping based on resource name (calico node name).
-			n := upd.KVPair.Key.(model.ResourceKey).Name
-			for kn, cn := range c.nodemapper {
-				if cn == n {
-					// Remove it from node map.
-					logrus.Debugf("Unmapping k8s node -> calico node. %s -> %s", kn, cn)
-					c.nodemapLock.Lock()
-					delete(c.nodemapper, kn)
-					c.nodemapLock.Unlock()
-					break
+			if c.config.SyncNodeLabels && c.config.DatastoreType != "kubernetes" {
+				for kn, cn := range c.nodemapper {
+					if cn == nodeName {
+						// Remove it from node map.
+						logrus.Debugf("Unmapping k8s node -> calico node. %s -> %s", kn, cn)
+						c.nodemapLock.Lock()
+						delete(c.nodemapper, kn)
+						c.nodemapLock.Unlock()
+						break
+					}
+				}
+			}
+
+			if c.config.AutoHostEndpoints == "enabled" {
+				hepName := c.generateAutoHostendpointName(nodeName)
+				err := c.deleteHostendpointWithRetries(hepName)
```

Comment:
We should probably only do this if we're in sync - if we get deletions during the sync phase then we'll handle them when we finally get in sync.

Reply:
2e72d9f